File: frontend/lib/services/page-analysis.ts
import { fetchAPI } from '@/lib/api-client';

// Page Metadata
export interface PageMetadata {
  title: string;
  description: string;
  author: string;
  keywords: string;
  canonical: string;
  robots: string;
}

// Content Analysis
export interface ContentAnalysis {
  word_count: number;
  sentence_count: number;
  avg_words_per_sentence: number;
  content_length: number;
}

// AEO Signals
export interface AEOSignals {
  has_faq_structure: boolean;
  has_how_to_structure: boolean;
  has_article_structure: boolean;
  question_content_ratio: number;
  has_clear_entity_references: boolean;
  conversational_readiness_score: number;
}

// Link Info
export interface LinkInfo {
  href: string;
  text: string;
  title: string;
  is_external: boolean;
  nofollow: boolean;
}

// Image Info
export interface ImageInfo {
  src: string;
  alt: string;
  has_alt: boolean;
}

// Page Analysis Response
export interface PageAnalysisResponse {
  id?: string;
  user_id?: string;
  brand_id?: string;
  url: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  metadata?: PageMetadata;
  content_analysis?: ContentAnalysis;
  seo_warnings: string[];
  headings: Record<string, string[]>;
  additional_tags: Record<string, any>;
  links: LinkInfo[];
  images: ImageInfo[];
  keywords: Record<string, number>;
  bigrams: Record<string, number>;
  trigrams: Record<string, number>;
  aeo_signals?: AEOSignals;
  content_hash?: string;
  created_at?: string;
  llm_analysis?: LLMAnalysisResponse;
  error?: string;
}

// N-E-E-A-T Scores
export interface NEEATScores {
  notability: number;
  experience: number;
  expertise: number;
  authority: number;
  trust: number;
  transparency: number;
}

// Entity Analysis
export interface EntityAnalysis {
  entity_assessment: string;
  knowledge_panel_readiness: number;
  key_improvements: string[];
}

// Credibility Analysis
export interface CredibilityAnalysis {
  credibility_assessment: string;
  neeat_scores?: NEEATScores;
  trust_signals: string[];
}

// Conversation Analysis
export interface ConversationAnalysis {
  conversation_readiness: string;
  query_patterns: string[];
  engagement_score: number;
  gaps: string[];
}

// Platform Presence
export interface PlatformPresence {
  platform_coverage: Record<string, string>;
  visibility_scores?: {
    search_engines: number;
    ai_assistants: number;
    knowledge_graphs: number;
    social: number;
  };
  optimization_opportunities: string[];
}

// LLM Analysis Scores
export interface LLMAnalysisScores {
  entity_score: number;
  credibility_score: number;
  conversation_score: number;
  platform_score: number;
  overall_score: number;
}

// LLM Analysis Response
export interface LLMAnalysisResponse {
  entity_analysis?: EntityAnalysis;
  credibility_analysis?: CredibilityAnalysis;
  conversation_analysis?: ConversationAnalysis;
  platform_presence?: PlatformPresence;
  scores?: LLMAnalysisScores;
  quick_wins: string[];
  strategic_recommendations: string[];
  error?: string;
}

// Website Crawl Keyword
export interface CrawlKeyword {
  word: string;
  count: number;
}

// Website Crawl Response
export interface WebsiteCrawlResponse {
  id?: string;
  user_id?: string;
  brand_id?: string;
  base_url: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  total_pages: number;
  pages: PageAnalysisResponse[];
  keywords: CrawlKeyword[];
  duplicate_pages: string[][];
  errors: string[];
  created_at?: string;
  completed_at?: string;
}

// Request interfaces
export interface PageAnalysisRequest {
  url: string;
  analyze_headings?: boolean;
  analyze_extra_tags?: boolean;
  extract_links?: boolean;
  run_llm_analysis?: boolean;
  llm_provider?: 'openai' | 'anthropic' | 'openrouter';
}

export interface WebsiteCrawlRequest {
  base_url: string;
  sitemap_url?: string;
  max_pages?: number;
  follow_links?: boolean;
  run_llm_analysis?: boolean;
  llm_provider?: 'openai' | 'anthropic' | 'openrouter';
}

// Page Analysis Service
export const pageAnalysisService = {
  /**
   * Analyze a single page for SEO/AEO signals
   */
  analyzePage: async (request: PageAnalysisRequest): Promise<PageAnalysisResponse> => {
    return fetchAPI<PageAnalysisResponse>('/page-analysis/analyze', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  },

  /**
   * Get a previously completed page analysis
   */
  getAnalysis: async (analysisId: string): Promise<PageAnalysisResponse> => {
    return fetchAPI<PageAnalysisResponse>(`/page-analysis/analyze/${analysisId}`);
  },

  /**
   * Start a website crawl
   */
  crawlWebsite: async (request: WebsiteCrawlRequest): Promise<WebsiteCrawlResponse> => {
    return fetchAPI<WebsiteCrawlResponse>('/page-analysis/crawl', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  },

  /**
   * Get crawl status and results
   */
  getCrawlStatus: async (crawlId: string): Promise<WebsiteCrawlResponse> => {
    return fetchAPI<WebsiteCrawlResponse>(`/page-analysis/crawl/${crawlId}`);
  },

  /**
   * Run LLM analysis on page data
   */
  runLLMAnalysis: async (
    pageData: Record<string, any>,
    provider: 'openai' | 'anthropic' | 'openrouter' = 'openai'
  ): Promise<LLMAnalysisResponse> => {
    return fetchAPI<LLMAnalysisResponse>(`/page-analysis/llm-analyze?provider=${provider}`, {
      method: 'POST',
      body: JSON.stringify(pageData),
    });
  },

  /**
   * Extract keywords from a URL
   */
  extractKeywords: async (url: string): Promise<{
    url: string;
    keywords: Record<string, number>;
    bigrams: Record<string, number>;
    trigrams: Record<string, number>;
    word_count: number;
  }> => {
    return fetchAPI(`/page-analysis/keywords/extract?url=${encodeURIComponent(url)}`);
  },
};

File: backend/app/models/page_analysis.py
"""
Page Analysis Models - Pydantic models for page and website SEO analysis.
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from uuid import UUID
from datetime import datetime
from enum import Enum


class PageAnalysisStatus(str, Enum):
    """Status of a page analysis."""
    pending = "pending"
    processing = "processing"
    completed = "completed"
    failed = "failed"


class PageMetadata(BaseModel):
    """Page metadata extracted during analysis."""
    title: str = ""
    description: str = ""
    author: str = ""
    keywords: str = ""
    canonical: str = ""
    robots: str = ""


class ContentAnalysis(BaseModel):
    """Content analysis metrics."""
    word_count: int = 0
    sentence_count: int = 0
    avg_words_per_sentence: float = 0
    content_length: int = 0


class AEOSignals(BaseModel):
    """AI Engine Optimization signals."""
    has_faq_structure: bool = False
    has_how_to_structure: bool = False
    has_article_structure: bool = False
    question_content_ratio: int = 0
    has_clear_entity_references: bool = False
    conversational_readiness_score: int = 0


class LinkInfo(BaseModel):
    """Information about a link on the page."""
    href: str
    text: str = ""
    title: str = ""
    is_external: bool = False
    nofollow: bool = False


class ImageInfo(BaseModel):
    """Information about an image on the page."""
    src: str
    alt: str = ""
    has_alt: bool = False


class PageAnalysisBase(BaseModel):
    """Base model for page analysis."""
    url: str
    status: PageAnalysisStatus = PageAnalysisStatus.pending
    metadata: Optional[PageMetadata] = None
    content_analysis: Optional[ContentAnalysis] = None
    seo_warnings: List[str] = []
    headings: Dict[str, List[str]] = {}
    additional_tags: Dict[str, Any] = {}
    links: List[LinkInfo] = []
    images: List[ImageInfo] = []
    keywords: Dict[str, int] = {}
    bigrams: Dict[str, int] = {}
    trigrams: Dict[str, int] = {}
    aeo_signals: Optional[AEOSignals] = None
    content_hash: Optional[str] = None


class PageAnalysisCreate(BaseModel):
    """Request model for creating a page analysis."""
    url: str
    analyze_headings: bool = True
    analyze_extra_tags: bool = True
    extract_links: bool = True
    run_llm_analysis: bool = False
    llm_provider: Optional[str] = "openai"


class PageAnalysisResponse(PageAnalysisBase):
    """Response model for page analysis."""
    id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    brand_id: Optional[UUID] = None
    created_at: Optional[datetime] = None
    llm_analysis: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    
    class Config:
        from_attributes = True


class WebsiteCrawlRequest(BaseModel):
    """Request model for website crawl."""
    base_url: str
    sitemap_url: Optional[str] = None
    max_pages: int = Field(default=50, le=200, ge=1)
    follow_links: bool = True
    run_llm_analysis: bool = False
    llm_provider: Optional[str] = "openai"


class CrawlKeyword(BaseModel):
    """Keyword found during crawl."""
    word: str
    count: int


class WebsiteCrawlResponse(BaseModel):
    """Response model for website crawl."""
    id: Optional[UUID] = None
    user_id: Optional[UUID] = None
    brand_id: Optional[UUID] = None
    base_url: str
    status: PageAnalysisStatus = PageAnalysisStatus.pending
    total_pages: int = 0
    pages: List[PageAnalysisResponse] = []
    keywords: List[CrawlKeyword] = []
    duplicate_pages: List[List[str]] = []
    errors: List[str] = []
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# LLM Analysis specific models
class NEEATScores(BaseModel):
    """N-E-E-A-T-T credibility scores."""
    notability: int = Field(default=0, ge=0, le=100)
    experience: int = Field(default=0, ge=0, le=100)
    expertise: int = Field(default=0, ge=0, le=100)
    authority: int = Field(default=0, ge=0, le=100)
    trust: int = Field(default=0, ge=0, le=100)
    transparency: int = Field(default=0, ge=0, le=100)


class EntityAnalysisResult(BaseModel):
    """Entity optimization analysis result."""
    entity_assessment: str = ""
    knowledge_panel_readiness: int = Field(default=0, ge=0, le=100)
    key_improvements: List[str] = []


class CredibilityAnalysisResult(BaseModel):
    """Credibility analysis result."""
    credibility_assessment: str = ""
    neeat_scores: Optional[NEEATScores] = None
    trust_signals: List[str] = []


class ConversationAnalysisResult(BaseModel):
    """Conversational readiness analysis result."""
    conversation_readiness: str = ""
    query_patterns: List[str] = []
    engagement_score: int = Field(default=0, ge=0, le=100)
    gaps: List[str] = []


class PlatformVisibilityScores(BaseModel):
    """Visibility scores per platform type."""
    search_engines: int = Field(default=0, ge=0, le=100)
    ai_assistants: int = Field(default=0, ge=0, le=100)
    knowledge_graphs: int = Field(default=0, ge=0, le=100)
    social: int = Field(default=0, ge=0, le=100)


class PlatformPresenceResult(BaseModel):
    """Platform presence analysis result."""
    platform_coverage: Dict[str, str] = {}
    visibility_scores: Optional[PlatformVisibilityScores] = None
    optimization_opportunities: List[str] = []


class LLMAnalysisScores(BaseModel):
    """Summary scores from LLM analysis."""
    entity_score: float = Field(default=0, ge=0, le=100)
    credibility_score: float = Field(default=0, ge=0, le=100)
    conversation_score: float = Field(default=0, ge=0, le=100)
    platform_score: float = Field(default=0, ge=0, le=100)
    overall_score: float = Field(default=0, ge=0, le=100)


class LLMAnalysisResponse(BaseModel):
    """Complete LLM analysis response."""
    entity_analysis: Optional[EntityAnalysisResult] = None
    credibility_analysis: Optional[CredibilityAnalysisResult] = None
    conversation_analysis: Optional[ConversationAnalysisResult] = None
    platform_presence: Optional[PlatformPresenceResult] = None
    scores: Optional[LLMAnalysisScores] = None
    quick_wins: List[str] = []
    strategic_recommendations: List[str] = []
